chapter02/worksheets/solution2.1.7.9_object_main.scala

```diff
@@ -0,0 +1,33 @@
+/** author: Andreas Röhler */
+/** author: Sergei Winitzki */ 
+
+/** 
+  Exercise 2.1.7.9
+  Given a Seq[(String, Int)] showing a list of purchased items (where item names
+may repeat), compute a Map[String, Int] showing the total counts. So, for the input:
+Seq(("apple", 2), ("pear", 3), ("apple", 5), ("lemon", 2), ("apple", 3))
+
+the output must be Map("apple" -> 10, "pear" -> 3, "lemon" -> 2).
+Hint: use groupBy, map, sum.
+ */
+
+object TotalCount {
+  def totalCount(a: Seq[(String, Int)]): Map[String, Int] = {
+    a.groupBy(s => s._1).map { case (x, y) => (x, y.map(_._2).sum) }
+  }
+  def main(args: Array[String]) {
```

Comment:
This object has a `main` function, which will make this program executable when compiled.

But there is no code that actually calls this `main` function. So, running this code in the tests will not fail even if there is an error.

I suggest adding (after line 26) the code `TotalCount.main(Array())` so that the code is actually run while running tests.

Reply:
Hmm, can do that. However, don't really understand. Because, when running it from shell, the if-clause inside main is executed, it prints "result" :

$ scala solution2.1.7.9_object_main.scala
result: Map(pear -> 3, apple -> 10, lemon -> 2)

Reply:
This is perhaps the behavior of the REPL. Let me look at the test logs to see if this is also happening in tests. But I wouldn't expect that to happen.

Reply:
I don't see anything printed: https://github.com/winitzki/sofp-solutions/actions/runs/7186906098/job/19573337500?pr=35

Reply:
Now, when running from bash, get a warning: 

solution2.1.7.9_object_main.scala:27: warning: Script has a main object but statement is disallowed
TotalCount.main(Array())
               ^
one warning found
result: Map(pear -> 3, apple -> 10, lemon -> 2)

Reply:
Okay, let's see if totalCount may be called from inside main.

Reply:
BTW, assume you don't see the printed, because checks are run on return-values, which may not consider printed stuff(?)

Reply:
All code is run, so print() functions are also run. You can see one print result in the test log.

You already call `totalCount` from inside `main`. But the code never calls `main`.

Reply:
After your changes, now I see in the test logs:
```
result: HashMap(apple -> 10, pear -> 3, lemon -> 2)
```

Reply:
Hmm, see that too, but only from 2.1.7.2 - not 2.1.7.9

 2023-12-13T10:24:13.0102667Z check_solution2_1_7_2_AR: this should be empty: Vector()result: HashMap(apple -> 10, pear -> 3, lemon -> 2)

Reply:
You see, 2.1.7.2 prints `2023-12-13T10:24:13.0102667Z check_solution2_1_7_2_AR: this should be empty: Vector()` but does not print a newline after that. Then 2.1.7.9 prints `result: HashMap(apple -> 10, pear -> 3, lemon -> 2)` in the same line.

Reply:
Okay, thanks. Will make the print-message a bit more explicit, so that's easier to spot.

Reply:
If you also add `\n` to the `printf` function in solution 2.1.7.2 then it will be better formatted and you don't need to add `\n` here.